### ISSUE TITLE
ManipulationHandler needs Translation Constraints

### ISSUE THREAD

**Issue by limbusdev** [microsoft/MixedRealityToolkit-Unity#4154]
## Overview

`ManipulationHandler` provides settings for "Constraints on Rotation" and "Constraints on Movement". The Options available however are not what one would expect.

"Constraints of Rotation" is fine:

+ "None"
+ "X-Axis Only"
+ "Y-Axis Only"
+ "Z-Axis Only"

"Constraints on Movement" are less intuitive:

+ "None"
+ "Fix Distance from Head"

## Expected Behavior

At least a third Option "Constraints on Translation" with Options:

+ "None"
+ "Along X-Axis Only"
+ "Along Y_Axis Only"
+ "Along Z-Axis Only"
+ "In X-Y-Plane Only"
+ "In X-Z-Plane Only"
+ "In Y-Z-Plane Only"
+ "All"

## Actual Behavior

The Option is missing

## Mixed Reality Toolkit Release Version

MRTK v2.0.0 RC1

## Workaround

As a Workaround I introduced a new **enum** `TranslationConstraintType`:

```cs
public enum TranslationConstraintType
        {
            None,
            XAxisOnly,
            YAxisOnly,
            ZAxisOnly,
            XYPlaneOnly,
            XZPlaneOnly,
            YZPlaneOnly,
            All
        }
```

I added a public property to change in the editor:

```cs
[SerializeField]
[Tooltip("Constrain translation")]
private TranslationConstraintType constraintOnTranslation = TranslationConstraintType.None;

public TranslationConstraintType ConstraintOnTranslation => constraintOnTranslation;
```

And made the logic respect it:

```cs
switch (ConstraintOnTranslation)
                {
                    case TranslationConstraintType.XAxisOnly:
                        targetPosition.y = hostTransform.position.y;
                        targetPosition.z = hostTransform.position.z;
                        break;
                    case TranslationConstraintType.YAxisOnly:
                        targetPosition.x = hostTransform.position.x;
                        targetPosition.z = hostTransform.position.z;
                        break;
                    case TranslationConstraintType.ZAxisOnly:
                        targetPosition.y = hostTransform.position.y;
                        targetPosition.x = hostTransform.position.x;
                        break;
                    case TranslationConstraintType.XYPlaneOnly:
                        targetPosition.z = hostTransform.position.z;
                        break;
                    case TranslationConstraintType.XZPlaneOnly:
                        targetPosition.y = hostTransform.position.y;
                        break;
                    case TranslationConstraintType.YZPlaneOnly:
                        targetPosition.x = hostTransform.position.x;
                        break;
                    case TranslationConstraintType.All:
                        targetPosition = hostTransform.position;
                        break;
                    default:
                        break;
                }
```

**Comment by limbusdev** [comment]
PS: Please mark it as feature request

**Comment by wiwei** [comment]
@julenka Any concerns about this proposal? Seems reasonable to me (also seems like a good candidate for something that anyone including community could just issue a PR for assuming you're good with this proposal)

**Comment by julenka** [comment]
No, I think it's a great idea, and limbusdev has something that seems very reasonable already done.

I would recommend the following:

"Constraints on Movement":
- "None"
- "Along X-Axis Only"
- "Along Y-Axis Only"
- "Along Z-Axis Only"
- "In X-Y-Plane Only"
- "In X-Z-Plane Only"
- "In Y-Z-Plane Only"
-  "Fixed Distance from Head"

There is still the question whether the X/Y/Z axes should be in local space or world space. My guess would be to make these in local space (host transform space), since objects might be arbitrarily rotated relative to the world origin (for example, a menu could be rotated to face you, so now movement about the X axis would be in world X which isn't necessarily right to left relative to you.

**Comment by julenka** [comment]
I'd be curious about the priority of this -- is it something blocking a partner from progressing, something that is not blocking but significantly impacting progress, or something that's more of a nice to have. For example, limbusdev provided a workaround above, which could be implemented either by directly changing ManipulationHandler, or extending / inheriting from ManipulationHandler.

To me it feels like a nice to have, but it is not yet something that is blocking people.

**Comment by xlfpx** [comment]
I was about to open a feature request for exactly this. I think there are plenty of use-cases where this is going to be needed (rather sooner than later) - such as for moving pieces on a chess board, machines on a factory floor, pictures on a wall, etc.
An advanced manipulation handler would work together with surface magnetism to allow an object to be moved while stuck to a surface. E.g. for moving trees on a landscape.

**Comment by wiwei** [comment]
@MenelvagorMilsom FYI

**Comment by gilbdev** [comment]
i would also need this :)
my current workaround is to use another component script that constraint the wanted axis (Y in my case) during `LateUpdate()`
didn't took too long, but if this is directly integrated in hte constrained on movement, it would be great.

**Comment by Naphier** [comment]
Just hopping in to the newer MRTK utils after getting our HL2s in and this was one of the first things I was looking for. Unfortunately the code base makes it nearly impossible to customize the ManipulationHandler. I ended up having to copy the existing manipulation handler. Then I had to make ManipulationMoveLogic, TwoHandRotateLogic, and TwoHandScaleLogic all public (they were internal and for some reason Unity wouldn't let me access them even though the namespace of my new ManipulationHandler was the same as MRTK's manipulation handler).
Then there were still a fair amount of code changes to make as I would have prefered to assign custom constraints, but the class was not conducive to that at all.

**Comment by MenelvagorMilsom** [comment]
@Naphier we recently added a new manipulator as an experimental feature: `ObjectManipulator`. This component should be more customizable than its predecessor `ManipulationHandler`. [These docs](https://microsoft.github.io/MixedRealityToolkit-Unity/Documentation/README_ObjectManipulator.html) show you how to set up a `ObjectManipulator` and how to customize constraints. This component will be supported in the future, whereas `ManipulationHandler` will be deprecated.

**Comment by Naphier** [comment]
Perfect. Exactly what I needed. Thanks!

**Comment by thalbern** [comment]
this is solved by using the new object manipulator in v2.4
